EIPS/eip-7377.md

```diff
@@ -0,0 +1,127 @@
+
+#### Processing
+
+Execution a MigrationTransaction transaction has two parts.
+
+##### Contract Deployment
+
+Unlike standard contract deployment, the MigrationTransaction directly specifies what `code` value of the sender's account should be set to.
+
+As the first step of processing the transaction, set the sender's `code` value equal to `state[tx.codePtr].code`. Next, iterate through each tuple in `tx.storage` and in the sender's storage trie, set `storage[t.first] = t.second`.
+
+##### Transaction Execution
+
+Now instantiate an EVM call into the sender's account using the same rules as [EIP-1559](./eip-1559) and set the transaction's origin to be `keccak256(sender)[0..20]`.
+
+## Rationale
+
+### No `to` address field
+
+This transaction is only good for one time use to migrate an EOA to a smart contract. It is designed to immediately call the deployed contract, which is at the sender's address, after deployment to allow the sender to do any kind of further processing.
+
+### Code pointer for deployment
+
+Naively, one could design the migration transaction to have a field `code` that is just bytes. The outcome of this is that there would be substantial duplication of code calldata, since many users will want to deploy the exact same thing (a wallet). Using a pointer instead acknowledges the overwhelming use case for the transaction type, and exploits it as an optimization.
+
+### Cheaper storage
+
+Sometimes it's better to lead with a carrot instead of a stick.
+
+### Intrinsic does not account for contract deployment
+
+This is takes advantage of the fact that clients tend to single, unique copies of code. Therefore, the only operation here is changing a pointer in the state trie to desired code's pointer.
+
+Additionally, the EOA already exists because it has enough balance for the MigrationTransaction to be considered valid. Therefore, we don't need to pay a premium for adding a new account into the state trie.
+
+### Manipulating transaction origin
+
+Many applications have a security check `caller == origin` to verify the caller is an EOA. This is done to "protect" assets. While it is usually more of a bandage than an actual fix, we attempt to placate these projects by modifying the origin of the transaction so the check will continue performing it's duty.
+
+### One-time migration
+
+There is no technical reason we couldn't allow EOAs to change their code at anytime with this transaction type. The only inhibitor at the moment is [EIP-3607](./eip-3607) which will cause migration transactions to be considered invalid if they come from an account with code already deployed. A functional reason for retaining this behavior though is that it makes it simpler to reason about contracts and their upgradability.
```

Comment:
```suggestion
While there is no technical reason we couldn't allow EOAs to submit more than one `MigrationTransaction`, [EIP-3607](./eip-3607.md) causes subsequent migration transactions to be invalid. Retaining this behavior makes it simpler to reason about contracts and their upgrade-ability.
```
```suggestion
There is no technical reason we couldn't allow EOAs to change their code at any time with this transaction type. The only inhibitor at the moment is [EIP-3607](./eip-3607.md) which will cause migration transactions to be considered invalid if they come from an account with code already deployed. A functional reason for retaining this behavior though is that it makes it simpler to reason about contracts and their upgrade-ability.
```
Fix three typos (any time; upgrade-ability; missing .md). Optionally rephrase this paragraph.

Reply:
```suggestion
There is no technical reason we couldn't allow EOAs to change their code at any time with this transaction type. The only inhibitor at the moment is [EIP-3607](./eip-3607) which will cause migration transactions to be considered invalid if they come from an account with code already deployed. A functional reason for retaining this behavior though is that it makes it simpler to reason about contracts and their upgradability.
```